docs/docs/protocol-specs/gas-and-fees/kernel-tracking.md

```diff
@@ -0,0 +1,523 @@
+    +CallRequest[MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX] public_call_stack
+}
+
+class CombinedConstantData {
+    +Header historical_header
+    +TxContext tx_context
+}
+CombinedConstantData --> Header
+CombinedConstantData --> TxContext
+
+class Header {
+    +GlobalVariables global_variables
+}
+Header --> GlobalVariables
+
+class GlobalVariables {
+    +GasFees gas_fees
+}
+GlobalVariables --> GasFees
+
+class GasFees {
+    +Fr fee_per_da_gas
+    +Fr fee_per_l2_gas
+}
+
+class TxContext {
+  +GasSettings gas_settings
+}
+TxContext --> GasSettings
+
+class GasSettings {
+    +Gas gas_limits
+    +Gas teardown_gas_limits
+    +GasFees max_fees_per_gas
+    +Field inclusion_fee
+}
+GasSettings --> Gas
+GasSettings --> GasFees
+
+class Gas {
+    +u32 l2_gas
+    +u32 da_gas
+}
+
+class TxRequest {
+  +TxContext tx_context
+}
+TxRequest --> TxContext
+
+class PrivateKernelInitCircuitPrivateInputs {
+    +PrivateCallData private_call_data
+    +TxRequest tx_request
+}
+PrivateKernelInitCircuitPrivateInputs --> PrivateCallData
+PrivateKernelInitCircuitPrivateInputs --> TxRequest
+
+class PrivateKernelInnerCircuitPrivateInputs {
+    +PrivateKernelData previous_kernel
+    +PrivateCallData private_call
+}
+PrivateKernelInnerCircuitPrivateInputs --> PrivateKernelData
+PrivateKernelInnerCircuitPrivateInputs --> PrivateCallData
+
+class PrivateKernelData {
+    +PrivateKernelCircuitPublicInputs public_inputs
+}
+PrivateKernelData --> PrivateKernelCircuitPublicInputs
+
+class PrivateKernelTailCircuitPrivateInputs {
+    +PrivateKernelData previous_kernel
+}
+PrivateKernelTailCircuitPrivateInputs --> PrivateKernelData
+
+class PrivateKernelTailToPublicCircuitPrivateInputs {
+    +PrivateKernelData previous_kernel
+}
+PrivateKernelTailToPublicCircuitPrivateInputs --> PrivateKernelData
+
+```
+
+## Private Context Initialization
+
+Whenever a private function is run, it has a `PrivateContext` associated with it, which is initialized in part from a `PrivateContextInputs` object. 
+
+The [gas settings that users specify](./specifying-gas-fee-info.md) become part of the values in the `TxContext` within the `PrivateContextInputs` of the [entrypoint](./tx-setup-and-teardown.md#defining-setup). These values are copied to the `PrivateCircuitPublicInputs`.
+
+The same `TxContext` is provided as part of the `TxRequest` in the `PrivateKernelInitCircuitPrivateInputs`. This is done to ensure that the `TxContext` in the `PrivateCallData` (what was executed) matches the `TxContext` in the `TxRequest` (users' intent).
+
+
+## Private Kernel Init
+
+The PrivateKernelInit circuit takes in a `PrivateCallData` and a `TxRequest` and outputs a `PrivateKernelCircuitPublicInputs`.
+
+It must:
+
+- check that the `TxContext` provided as in the `TxRequest` input matches the `TxContext` in the `PrivateCallData` 
+- check that the max fees per gas specified by the user are greater than the header's fee per gas for both DA and L2
```

Comment:
Got this one late: I'm not sure we can check this here, since the fee per gas for the block is unknown in private-land.